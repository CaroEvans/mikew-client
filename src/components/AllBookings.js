import React from 'react'
import moment from 'moment'
import Booking from './Booking'
import Loading from './Loading'
import { api } from '../api/init';

const jwtDecode = require('jwt-decode')

class AllBookings extends React.Component {
    state = { 
        declined: [],
        approved: [],
        pending: [],
        completed: [],
        cancelled: [],
        loading: false,
        show: "pending"
     }

     isAdmin = () =>{ //used in callbacks, and in statements, to check whether the decoded token is an admin.
        const decoded = jwtDecode(localStorage.getItem('token'))
        return decoded.role === "admin" 
     }

    componentDidMount(){
        this.setState({loading: true}) //displays loading screen
        const decoded = jwtDecode(localStorage.getItem('token'))
        if( this.isAdmin()){
            this.getAdminBookings()
        }else{ this.getUserBookings(decoded.sub) }
    }

    getAdminBookings = () => {
        api.get("bookings") //returns ALL bookings, not limited to user ID.
        .then((response) => {
            response.data.forEach((data) => { this.checkBookingStatus(data) })
            this.setState({loading: false }) })
        .catch((err) => { console.log(err);this.setState({loading: false }) })
    }

    getUserBookings = (id) => {
        api.get("users/bookings",{params : {id: id }})
        .then((response) => {
            response.data.forEach((data) => { this.checkBookingStatus(data) })
            this.setState({ loading: false }) })
        .catch((err) => { console.log(err) })
    }

    checkBookingStatus = (booking) => { //decides which category the bookings go into.
        let bookingStatus = booking.bookingStatus //booking.bookingsStatus is either Pending, Approved, Cancelled, Completed.
        this.setState({ [bookingStatus]: [...this.state[bookingStatus], booking] })
    }
        
    bookingSection = (bookingArr, admin) =>{
        return bookingArr.map((booking) => {
            return (
                <div className="contact" key = {booking._id}>  
                        <Booking {...booking}/>
                        {/* check for related buttons to display*/}
                        {admin && booking.bookingStatus === 'pending' ? 
                        <button className="approve-button" onClick={() => 
                            this.handleBookingChange(booking._id,booking.bookingStatus, 'approved')
                        }> Approve Booking? </button>:<p></p>}

                        {booking.bookingStatus === 'approved' || booking.bookingStatus === 'pending' ?
                        <button onClick={() => 
                            this.handleBookingChange(booking._id,booking.bookingStatus,'cancelled' )
                        }> Cancel Booking? </button>:<p></p>
                    }
                </div>
            )
        })
    }

    //Handles which booking category is shown, called on button click.
    toggleShow = (section) => {
        this.setState({show: section})
    }

        // updates state and back end when status of booking is changed
        handleBookingChange = (bookingId, current, update) =>{

            let currentBookings = this.state[current]
            let updateBookings = this.state[update]

            api.put('bookings/id', {id: bookingId , bookingStatus: `${update}`})
            .then((response) => { 
                currentBookings.forEach((obj) => {
                    if(obj._id === bookingId){
                        // find bookings updates status and adds in to new array
                        obj.bookingStatus = update
                        updateBookings.push(obj)
                        // deletes from current state array
                        let index = currentBookings.indexOf(obj)
                        currentBookings.splice(index,1)

                        this.setState({
                            [update] : updateBookings,
                            [current]: currentBookings
                        })
                    }
                })
            }).catch((err) => {console.log(err)})
        }
    
        // not being used right now 
        readableDate = (date) => ( moment(date, 'YYYYMMDD').format('MMM Do YY') )

    render() { 
        const { completed, approved, pending, cancelled } = this.state
        const { bookingSection, isAdmin, toggleShow } = this
        return ( 
            <div className="allbookings"> 
                <button onClick={() => toggleShow("pending")}> Pending </button>
                <button onClick={() => toggleShow("approved")}> Approved </button>
                <button onClick={() => toggleShow("cancelled")}> Cancelled </button>
                <button onClick={() => toggleShow("completed")}> Completed </button>
                
                {/* depending on what this.state.show is, is what category of bookings
                will be displayed, so only one can be shown at a time */}
                {(this.state.show === "pending") && 
                <div>
                    {bookingSection(pending,isAdmin())}
                </div>
                }
                {(this.state.show === "approved") && 
                <div>
                    {bookingSection(approved,isAdmin())}
                </div>
                }
                {(this.state.show === "completed") && 
                <div>
                    {bookingSection(completed,isAdmin())}
                </div>
                }
                {(this.state.show === "cancelled") && 
                <div >
                    {bookingSection(cancelled,isAdmin())}
                </div>
                }

            </div>
         )
    }
}
 
export default AllBookings;