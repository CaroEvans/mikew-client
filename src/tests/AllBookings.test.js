import React from 'react'
import AllBookings from '../components/AllBookings.js'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import './setUpTests'

describe('AllBookings Component', () => {
  it('matches the snapshot', () => {
      const tree = renderer.create(<AllBookings />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  it('shows all bookings', () => {
      const wrapper = shallow(<AllBookings />)
      const bookingsState = wrapper.state().bookings
      expect(bookingsState).toEqual([])
  })
  it('shows Loading Component', () => {
    const wrapper = shallow(<AllBookings />)
    const text = wrapper.find('.loadingScreen').text()
    expect(text).toContain('<Loading />')
  })
})