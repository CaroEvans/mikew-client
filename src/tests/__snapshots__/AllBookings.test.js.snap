// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllBookings Component matches the snapshot 1`] = `
<div
  className="Loading"
>
  <h1>
     Loading... 
  </h1>
  <img
    src="http://i79.photobucket.com/albums/j157/mdcummings1/guitar.gif"
  />
</div>
`;
